server/service/src/document/document_service.rs

```diff
@@ -0,0 +1,440 @@
+use chrono::Utc;
+
+use jsonschema::JSONSchema;
+use repository::{
+    AncestorDetail, Document, DocumentFilter, DocumentRepository, EqualFilter,
+    JsonSchemaRepository, RepositoryError, StorageConnection,
+};
+
+use crate::service_provider::ServiceContext;
+
+use super::{
+    common_ancestor::{common_ancestors, AncestorDB, CommonAncestorError, InMemoryAncestorDB},
+    merge::{three_way_merge, two_way_merge, TakeLatestConflictSolver},
+    raw_document::RawDocument,
+    topological_sort::{extract_tree, topo_sort},
+};
+
+#[derive(Debug)]
+pub enum DocumentInsertError {
+    /// Input document doesn't match the provided json schema
+    InvalidDataSchema(Vec<String>),
+    /// Document version needs to be merged first. Contains an automerged document which can be
+    /// reviewed and/or inserted.
+    /// If no automerged document is provided the document couldn't be automerged, e.g because the
+    /// merged data has an invalid data schema.
+    MergeRequired(Option<RawDocument>),
+    InvalidDocumentHistory,
+    /// Unable to finalise a document (assign an id)
+    FinalisationError(String),
+    DatabaseError(RepositoryError),
+    InternalError(String),
+}
+
+#[derive(Debug)]
+pub enum DocumentHistoryError {
+    DatabaseError(RepositoryError),
+    GraphError(String),
+}
+
+impl From<RepositoryError> for DocumentHistoryError {
+    fn from(err: RepositoryError) -> Self {
+        DocumentHistoryError::DatabaseError(err)
+    }
+}
+
+pub trait DocumentServiceTrait: Sync + Send {
+    fn get_document(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        name: &str,
+    ) -> Result<Option<Document>, RepositoryError> {
+        DocumentRepository::new(&ctx.connection).find_one_by_name(store_id, name)
+    }
+
+    fn get_documents(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        filter: Option<DocumentFilter>,
+    ) -> Result<Vec<Document>, RepositoryError> {
+        let filter = filter.map(|mut f| {
+            f.store_id = Some(EqualFilter::equal_to(store_id));
+            f
+        });
+        DocumentRepository::new(&ctx.connection).query(filter)
+    }
+
+    fn get_document_history(
```

Comment:
We should aim to provide front end consumable shapes, for history we would want to know what's changed, where and by whom. 

If we store field changes in document this becomes very easy:
```rust
let documents = /* get all documents for name */ 
let ordered_documents = /* order by timestamp */
/* I think that's it, easy for back end and front end */
```